src/test/scala/DSATest.scala

```diff
@@ -0,0 +1,707 @@
+    val results = RunUtils.loadAndTranslate(
+      BASILConfig(
+        loading = ILLoadingConfig(
+          inputFile = "examples/interproc_pointer_arithmetic/interproc_pointer_arithmetic.adt",
+          relfFile = "examples/interproc_pointer_arithmetic/interproc_pointer_arithmetic.relf",
+          specFile = None,
+          dumpIL = None,
+        ),
+        staticAnalysis = Some(StaticAnalysisConfig()),
+        boogieTranslation = BoogieGeneratorConfig(),
+        outputPrefix = "boogie_out",
+      )
+    )
+    val program = results.ir.program
+    val dsg = results.analysis.get.locals.get(program.procs("callee"))
+    val stack8 = dsg.adjust(dsg.find(dsg.stackMapping(8).cells(0)).getPointee)
+    val stack24 = dsg.adjust(dsg.find(dsg.stackMapping(24).cells(0)).getPointee)
+
+    // stack8 points to the formal argument object
+    assert(stack8.equals(dsg.adjust(dsg.formals(R0))))
+    assert(stack8.offset == 0)
+    // stack 24 points to the formal argument object at offset 16, instead
+    assert(stack24.equals(dsg.adjust(dsg.formals(R0)).node.get.cells(16)))
+  }
+
+
+  test("internal merge") {
+    // this is an internal merge (two cells of the same node overlap and are merged together)
+    val mem = SharedMemory("mem", 10000, 10000)
+    val locAssign1 = Assign(R6, BinaryExpr(BVADD, R0, BitVecLiteral(4, 64)), Some("00001"))
+    val locAssign2 = Assign(R7, BinaryExpr(BVADD, R0, BitVecLiteral(5, 64)), Some("00002"))
+    var program = prog(
+      proc("main",
+        block("operations",
+          locAssign1, // R6 = R0 + 4
+          locAssign2, // R7 = R0 + 5
+          MemoryAssign(mem,  R7, R1, BigEndian, 64, Some("00003")),
+          MemoryAssign(mem,  R6, R2, BigEndian, 64, Some("00004")),
+          ret
+        )
+      )
+    )
+
+    val returnUnifier = ConvertToSingleProcedureReturn()
+    program = returnUnifier.visitProgram(program)
+    val results = RunUtils.staticAnalysis(StaticAnalysisConfig(None, None, None), IRContext(Set.empty, Set.empty, Set.empty, Map.empty, Specification(Set(), Set(), Map(), List(), List(), List(), Set()), program))
+    val dsg: DSG = results.locals.get(program.mainProcedure)
+
+    // object of formals R1 and R2 are written to overlapping fields of the same node? causing them to be merged together
+    assert(dsg.adjust(dsg.formals(R1)).equals(dsg.adjust(dsg.formals(R2))))
+
+    // R6 and R7 address the same cell
+    assert(dsg.find(dsg.varToCell(locAssign1)(R6)).cell.equals(dsg.find(dsg.varToCell(locAssign2)(R7)).cell))
+    // however, they address different internal offets in those cells
+    assert(dsg.find(dsg.varToCell(locAssign1)(R6)).internalOffset == 0)
+    assert(dsg.find(dsg.varToCell(locAssign2)(R7)).internalOffset == 1)
+    // R6 points to input R1
+    assert(dsg.adjust(dsg.varToCell(locAssign1)(R6))._pointee.isDefined)
+    assert(dsg.adjust(dsg.adjust(dsg.varToCell(locAssign1)(R6)).getPointee).equals(dsg.adjust(dsg.formals(R1))))
```

Comment:
Is this because the analysis has no concept of most of the value of R1 (at address R0 + 5) being overwritten by the write of the value of R2 to address R0 + 4? The analysis just says that both R1 and R2 were written to R0 + 4 at some point? If this is the case, it should be explicit in the comments.

Reply:
the test comments says
`// object of formals R1 and R2 are written to overlapping fields of the same node? causing them to be merged together assert(dsg.adjust(dsg.formals(R1)).equals(dsg.adjust(dsg.formals(R2))))`
*R7 := R1 // equivalent to *(R6 + 1) := R1
*R6 := R2 
Since R6 and R7 point to the overlapping cells from the same node (Node pointed to offset 0 by R0) at offsets 4 and 5 respectively, the overlapping cells are merged together to form a cell encompassing both. This will need to point to two cells (the outgoing edges of cells at offset 4 and 5 respectively). So these two outgoing edges are unified. 
After the analysis is done, R6 and R7  both point to the same cell at different internal offsets, that cell will have one outgoing edge. Outgoing edge is a cell in the node resulting from the unification of the nodes of R1 and R2.

The test checks that R1 and R2 are pointing to the same cell since they have been unified and states that in comment.
it checks R6 and R7 are the slices with the same cell and different internal offsets.
then it checks that R6 points to R1
this is the same as R7 points to R2
or R6 points to R2
or R7 points to R1

Reply:
It would be helpful if this sort of thing, explaining what you are actually trying to test, was more explicit in the test comments - 'R6 points to input R1' is just not very clear on its own, and talking about something pointing to a register is also unintuitive.